### ISSUE TITLE
refineTagsAndSort can't work due to sequence length

### ISSUE THREAD

**Issue by hiro-aoki-mriid** [milaboratory/mixcr#829]
I am trying to run mixcr on single-cell TCRseq dataset obtained by BD Rhapsody.
However, "refineTagsAndSort" command did not work.

I got the following error massage. The cause of the error seems to be that the length of Cell Barcode sequnece of the library is longer than 30 nucleotides.
![image](https://user-images.githubusercontent.com/99735120/197316662-3464689f-7536-455c-bf6c-ad6646858227.png)

Cell Barcode of BD Rhapsody is 35 bps, including 4bp x2 linker sequences.
Structure of Cell Barcode and UMI are as follows. 
![image](https://user-images.githubusercontent.com/99735120/197316473-3a9c1a55-60fc-4b9e-96fe-0aa4cfc39cbe.png)

How can I apply "refineTagsAndSort" to the Cell Barcodes of BD Rhapsody data?
Or can I run the pipeline for single-cell TCRseq data with skipping "refineTagsAndSort" command?

Best,
Hiroyasu Aoki

**Comment by mizraelson** [comment]
Hi,
Can you also provide the exact MiXCR align command that you used before refineAndSortTags?

**Comment by hiro-aoki-mriid** [comment]
Hi,
MiXCR align command I used is as follows;

mixcr align \
    --preset 10x-vdj-tcr \
    --species mmu \
    --tag-pattern "^N{0:3}(CELL:N{9}GTGAN{9}GACAN{9})(UMI:N{8})\^(R2:*)"　\
    --threads 24 \
    --force-overwrite \
      $file1.fastq.gz \
      $file2.fastq.gz \
      $samplename.vdjca

**Comment by mizraelson** [comment]
Hi, so the issue with the barcode can be solved by introducing several Cell barcodes instead of the single long one, e.g.

```
--tag-pattern "^N{0:3} (CELL1:N{9})gtca(CELL2:N{9})gaca
(CELL3:N{9})(UMI:N{8})\^(R2:*)"　
```

I would also use lowercase letters for the known spacers to allow fuzzy matching.

This pattern will require some adjustments in tag correction and assemble. 
I can get back to you in a couple of days with the specific preset for bd-rhapsody, which you will be able to use without specifying any additional parameters (except for the `--species` obviously). I just have to test in on publicly available Rhapsody raw dataset.

**Comment by hiro-aoki-mriid** [comment]
Thank you for your kind response!

As per your instruction, I ran **align** again with the following --tag-pattern parameter.
`--tag-pattern "^N{0:3}(CELL1:N{9})gtga(CELL2:N{9})gaca(CELL3:N{9})(UMI:N{8})\^(R2:*)"`

I also changed the whitelist of cell barcode in **refineAndSortTags** to exclude 4bp x2 Linker Sequence (35 bp → 27 bp).

However, I faced the error again as below.
![image](https://user-images.githubusercontent.com/99735120/197525587-d51913d3-1f9f-4ab5-8645-780b3992cb83.png)

Could you give me any advice to resolve this problem?

Best,

**Comment by mizraelson** [comment]
Hi, we are currently working on the specific preset for BD Rhapsody data. It will include a lot of protocol specific features. As soon as it's done, I will get back to you with the MiXCR version which supports it.

**Comment by hiro-aoki-mriid** [comment]
Hi, thank you so much for your responses.
I understand the situation and will wait until the new version of MiXCR is released.
Best,

**Comment by alefrol638** [comment]
Hi, I saw that you have merged the „protocols“ pull request with the dev branch, which contained the bd rhapsody preset option, but couldn’t find it in your documentation. 

Is it now possible to use mixcr on bdrhaspody data and could you maybe send me a small example code snippet, if yes? 

Thank you for constantly developing such a great tool and taking time for me!

**Comment by dbolotin** [comment]
Hi, we just released [mixcr 4.1.1](https://github.com/milaboratory/mixcr/releases/tag/v4.1.1) please see all the presets for BD Rhapsody in the [corresponding docs section](https://docs.milaboratories.com/mixcr/reference/overview-built-in-presets/#bd-rhapsody).
